client/thin-replica-client/src/replica_state_snapshot_client.cpp

```diff
@@ -21,20 +21,28 @@ using concord::client::concordclient::OutOfRangeSubscriptionRequest;
 using concord::client::concordclient::InternalError;
 using concord::client::concordclient::EndOfStream;
 using concord::client::concordclient::StreamUnavailable;
+using concord::client::concordclient::RequestOverload;
 using concord::client::concordclient::SnapshotQueue;
 using vmware::concord::replicastatesnapshot::StreamSnapshotRequest;
 using vmware::concord::replicastatesnapshot::StreamSnapshotResponse;
 
 namespace client::replica_state_snapshot_client {
 void ReplicaStateSnapshotClient::readSnapshotStream(const SnapshotRequest& request,
                                                     std::shared_ptr<SnapshotQueue> remote_queue) {
+  if (count_of_concurrent_request_.load() > config_->concurrency_level) {
+    remote_queue->setException(std::make_exception_ptr(RequestOverload()));
+    return;
+  }
+  ++count_of_concurrent_request_;
```

Comment:
Thanks, @dartdart26 for this comment. I thought about it and due to the nice abstraction and design of our ThreadPool, it's not a problem. Let me explain.
As we know that our ThreadPool has a task queue. This task queue is such that it holds a task and schedules as soon as one of the threads is free. So if I keep the thread pool concurrency as 16, I am seeing that it's able to take almost 64 connections gracefully. So a factor of almost 4 times the number of available threads can be taken care of due to the ThreadPool. But when we bombard the thread pool with 1000 or more threads we see throttling in client service. 
Now if we take the worst-case count, we will see that it's 2*concurrency_level and practically that is acceptable. This is also a good thing since it will allow us to use the full potential of the ThreadPool. So I used it as a loose counter here. So the clipping is bounded between [concurrency_level, 2* concurrency_level] which will utilize the resources and if it cannot, then the server will respond gracefully.

After this change, I will test this with a 10K and 100K parallel connection and if I see the throttling then I will use the strict counter for this purpose. But I will need to do that in a reservation host (or in a castor host) as my machine will die before 10K parallel connections.